visualizer/semanticsHelper.js

```diff
@@ -201,21 +201,44 @@
     return defaults;
   }
 
+  // Merge the actions in the `optActionDict` into `actionDict`
+  function mergeActionDict(name, actionDict, optActionDict) {
+    if (!optActionDict) {
+      return;
+    }
+    if (optActionDict) {
+      Object.keys(optActionDict).forEach(function(key) {
+        if (actionDict[key]) {
+          return;
+        }
+        var actionArguments = optActionDict[key]._actionArguments;
+        var actionBody = optActionDict[key]._actionBody;
+        actionDict[key] = wrapAction(name, actionArguments, actionBody);
+      });
+    }
+  }
+
   // Add new operation/attribute to the semantics
-  function addSemanticOperation(type, name, optArguments) {
+  function addSemanticOperation(type, name, optArguments, optOrigActionDict) {
     var signature = name;
     if (type === 'Operation' && optArguments) {
       var argumentNames = Object.keys(optArguments);
       if (argumentNames.length > 0) {
         signature += '(' + argumentNames.join(',') + ')';
       }
     }
+
     semantics['add' + type](signature, initActionDict(type, name));
+    mergeActionDict(name, semantics._getActionDict(name), optOrigActionDict);
   }
-  ohmEditor.semantics.addListener('add:semanticOperation', function(type, name, optArguments) {
+  ohmEditor.semantics.addListener('add:semanticOperation', function(type, name, optArguments,
+      origActionDict) {
+    if (!/^[a-zA-Z_$][0-9a-zA-Z_$]*$/.test(name)) {
```

Comment:
Again, I would factor this check out into a separate function (e.g. `isOperationNameValid`), and a comment that explains what it should be checking (and why), and maybe a TODO to make it less restrictive in the future.